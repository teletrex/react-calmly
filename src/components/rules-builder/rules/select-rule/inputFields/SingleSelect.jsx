/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import find from 'lodash/find';

import { Dropdown} from '@carbon/react';
 import { useTranslation } from '../../../../../translation';
import RulesBuilderContext from '../../../RulesBuilderContext';
import { PROPTYPES, VALUES } from '../../constants';
import { getPrefixedClasses } from '../../helpers';

const SingleSelect = ({ id, isReadOnly, onValueChange, options, placeholder, value }) => {
  const { t } = useTranslation();
  const { isLight } = useContext(RulesBuilderContext);

  const selectedItem = find(options, { value }) || null;

  const handleOnChange = ({ selectedItem: newItem }) => {
    if (newItem.value !== value) {
      onValueChange(newItem.value);
    }
  };

  if (isReadOnly) {
    return (
      <span className={getPrefixedClasses(['--rule-value-readonly', '--text-space-left'])}>
        {selectedItem?.label}
      </span>
    );
  }

  return (
    <Dropdown
      ariaLabel={t('Select dropdown')}
      id={id}
      isInline
      isLight={isLight}
      items={options}
      onChange={handleOnChange}
      placeholder={placeholder || t('Choose an option')}
      selectedItem={selectedItem}
      withSearch
    />
  );
};

SingleSelect.propTypes = {
  id: PropTypes.string.isRequired,
  isReadOnly: PropTypes.bool,
  onValueChange: PropTypes.func.isRequired,
  options: PROPTYPES.SELECT_OPTIONS,
  placeholder: PropTypes.string,
  value: PropTypes.string,
};

SingleSelect.defaultProps = {
  isReadOnly: false,
  options: [],
  placeholder: '',
  value: VALUES.UNPROVIDED,
};

export default SingleSelect;
