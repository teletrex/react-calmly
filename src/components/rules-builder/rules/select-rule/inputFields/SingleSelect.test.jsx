/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React from 'react';
import { render, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import SingleSelect from './SingleSelect';

const options = [
  { label: 'Option 1', value: 'OPTION_1' },
  { label: 'Option 2', value: 'OPTION_2' },
];
const initialValue = options[0].value;

const defaultProps = {
  id: 'TestSelectRuleElement',
  isReadOnly: false,
  options,
  onValueChange: jest.fn(),
};

describe('SingleSelect', () => {
  beforeEach(() => {
    defaultProps.onValueChange.mockReset();
  });

  it('should render correctly', () => {
    const { container } = render(<SingleSelect {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with a provided value', () => {
    const { container } = render(<SingleSelect {...defaultProps} value={initialValue} />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly for read-only', () => {
    const { container, getByText } = render(
      <SingleSelect {...defaultProps} isReadOnly value={initialValue} />
    );
    expect(container).toMatchSnapshot();
    expect(getByText(options[0].label)).toBeInTheDocument();
  });

  it('should set the rule value when a value is entered', () => {
    const { getByRole, getByText } = render(<SingleSelect {...defaultProps} />);

    // expand the value dropdown:
    const dropdown = getByRole('listbox');
    const dropdownButton = within(dropdown).getByRole('button');
    userEvent.click(dropdownButton);

    // select a new value:
    userEvent.click(getByText(options[1].label));

    expect(defaultProps.onValueChange).toHaveBeenCalledWith(options[1].value);
  });

  it('should display the default placeholder', () => {
    const { getByText } = render(<SingleSelect {...defaultProps} />);

    expect(getByText('Choose an option')).toBeInTheDocument();
  });

  it('should display the provided placeholder', () => {
    const { getByText } = render(<SingleSelect {...defaultProps} placeholder="Test dropdown" />);

    expect(getByText('Test dropdown')).toBeInTheDocument();
  });
});
