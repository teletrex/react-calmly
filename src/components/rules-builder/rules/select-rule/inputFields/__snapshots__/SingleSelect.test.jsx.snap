// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleSelect should render correctly 1`] = `
<div>
  <div
    class="bx-proposal-dropdown-wrapper bx-proposal-dropdown-wrapper--inline"
  >
    <div
      class="bx-proposal-dropdown"
      role="listbox"
    >
      <span>
        <div
          aria-label="Choose an option"
          class="bx-proposal-dropdown__field"
          role="button"
          tabindex="0"
        >
          <span
            class="bx-proposal-dropdown__field__label"
          >
            Choose an option
          </span>
          <div
            class="bx-proposal-dropdown__field__menu-icon"
          >
            <svg
              aria-hidden="true"
              focusable="false"
              height="16"
              preserveAspectRatio="xMidYMid meet"
              style="will-change: transform;"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 11L3 6l.7-.7L8 9.6l4.3-4.3.7.7z"
              />
            </svg>
          </div>
        </div>
      </span>
    </div>
  </div>
</div>
`;

exports[`SingleSelect should render correctly for read-only 1`] = `
<div>
  <span
    class="bx--rule-value-readonly bx--text-space-left"
  >
    Option 1
  </span>
</div>
`;

exports[`SingleSelect should render correctly with a provided value 1`] = `
<div>
  <div
    class="bx-proposal-dropdown-wrapper bx-proposal-dropdown-wrapper--inline"
  >
    <div
      class="bx-proposal-dropdown"
      role="listbox"
    >
      <span>
        <div
          aria-label="Choose an option"
          class="bx-proposal-dropdown__field"
          role="button"
          tabindex="0"
        >
          <span
            class="bx-proposal-dropdown__field__label"
          >
            Option 1
          </span>
          <div
            class="bx-proposal-dropdown__field__menu-icon"
          >
            <svg
              aria-hidden="true"
              focusable="false"
              height="16"
              preserveAspectRatio="xMidYMid meet"
              style="will-change: transform;"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 11L3 6l.7-.7L8 9.6l4.3-4.3.7.7z"
              />
            </svg>
          </div>
        </div>
      </span>
    </div>
  </div>
</div>
`;
